cmd/kops/get_instances.go

```diff
@@ -0,0 +1,242 @@
+	"k8s.io/kops/util/pkg/tables"
+
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/resources/aws"
+
+	"github.com/spf13/cobra"
+	"k8s.io/kops/cmd/kops/util"
+	"k8s.io/kops/upup/pkg/fi/cloudup"
+
+	osCloudup "k8s.io/kops/upup/pkg/fi/cloudup/openstack"
+)
+
+func NewCmdGetInstances(f *util.Factory, out io.Writer, options *GetOptions) *cobra.Command {
+	getInstancesShort := i18n.T(`Display cluster instances.`)
+
+	getInstancesLong := templates.LongDesc(i18n.T(`
+	Display cluster instances.`))
+
+	getInstancesExample := templates.Examples(i18n.T(`
+	# Display all instances.
+	kops get instances
+	`))
+
+	cmd := &cobra.Command{
+		Use:     "instances",
+		Short:   getInstancesShort,
+		Long:    getInstancesLong,
+		Example: getInstancesExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx := context.TODO()
+
+			if err := rootCommand.ProcessArgs(args); err != nil {
+				exitWithError(err)
+			}
+
+			err := RunGetInstances(ctx, f, out, options)
+			if err != nil {
+				exitWithError(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func RunGetInstances(ctx context.Context, f *util.Factory, out io.Writer, options *GetOptions) error {
+
+	clientset, err := f.Clientset()
+	if err != nil {
+		return err
+	}
+
+	clusterName := rootCommand.ClusterName()
+	options.clusterName = clusterName
+	if clusterName == "" {
+		return fmt.Errorf("--name is required")
+	}
+
+	cluster, err := clientset.GetCluster(ctx, options.clusterName)
+	if err != nil {
+		return err
+	}
+
+	if cluster == nil {
+		return fmt.Errorf("cluster not found %q", options.clusterName)
+	}
+
+	cloud, err := cloudup.BuildCloud(cluster)
+	if err != nil {
+		return err
+	}
+
+	k8sClient, err := createK8sClient(cluster)
+	if err != nil {
+		return err
+	}
+
+	var status map[string]string
+	nodeList, err := k8sClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.V(2).Infof("error listing nodes: %v", err)
+	} else {
+		status, _ = getNodeStatus(ctx, cloud, clientset, cluster, nodeList.Items)
+	}
+
+	var instances []*resources.Instance
+
+	switch cloud.ProviderID() {
+	case kops.CloudProviderAWS:
+		rs, _ := aws.ListInstances(cloud, options.clusterName)
+		for _, r := range rs {
+			instances = append(instances, aws.GetInstanceFromResource(r))
+		}
+	case kops.CloudProviderOpenstack:
+		rs, _ := openstack.ListResources(cloud.(osCloudup.OpenstackCloud), options.clusterName)
+		for _, r := range rs {
+			if r.Type == "Instance" {
+				instances = append(instances, openstack.GetInstanceFromResource(r))
+			}
+		}
+	default:
+		return fmt.Errorf("cloud not supported")
+	}
+
+	switch options.output {
+	case OutputTable:
+		return instanceOutputTable(instances, status, out)
+	default:
+		return fmt.Errorf("Unsupported output format: %q", options.output)
+	}
+}
+
+func instanceOutputTable(instances []*resources.Instance, status map[string]string, out io.Writer) error {
+	t := &tables.Table{}
+	t.AddColumn("NAME", func(i *resources.Instance) string {
+		return i.Name
+	})
```

Comment:
In my case, all instances of the same IG are named the same. Easier than just having an IG column instead?
Also, maybe order should be consistent with table columns order.

Reply:
In AWS you typically have the same name for all nodes, but OS does not. So there are some value in the name column there. IG could be useful, but not that easy to get to. Could perhaps add it to `resources.Instance`. Easy enough to extract it from tags there.

Reply:
Let's leave as is. Thanks.